File: AndroidApp/AndroidQuizApp/app/src/main/java/com/example/androidquizapp/QuestionsRepository.kt
package com.example.androidquizapp

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.firebase.firestore.FirebaseFirestore
import com.example.androidquizapp.QuizViewModel.Question
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.QuerySnapshot




class QuestionsRepository {
    private val db = FirebaseFirestore.getInstance()
    private val _questions = MutableLiveData<List<Question>>()
    val questions: LiveData<List<Question>> = _questions
    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading

    init {
        fetchQuestions()
    }

    private fun fetchQuestions() {
        _isLoading.value = true
        val infoRef = db.collection("Quiz").document("Coding_questions_info")
        infoRef.get().addOnSuccessListener { infoSnap ->
            val totalQuestions = infoSnap.getLong("totalQuestions")?.toInt() ?: 0
            Log.d("TotalQuestions", totalQuestions.toString()) // Log totalQuestions value
            val numQuestions = 10
            val randomIds = getRandomIds(numQuestions, totalQuestions)
            Log.d("RandomIds", randomIds.toString()) // Log the generated random IDs
            val fetchedQuestions = mutableListOf<QuizViewModel.Question>()

            if (randomIds.isNotEmpty()) {
                val fetchPromises = randomIds.map { id ->
                    db.collection("Quiz").document(id.toString()).collection("Coding_questions").get()
                }

                Tasks.whenAllSuccess<QuerySnapshot>(fetchPromises)
                    .addOnSuccessListener { querySnapshots ->
                        querySnapshots.forEach { querySnapshot ->
                            querySnapshot.documents.forEach { doc ->
                                val question = QuizViewModel.Question.fromFirestoreData(
                                    doc.getLong("id")?.toInt() ?: 0, doc.data as Map<String, Any>
                                )
                                fetchedQuestions.add(question)
                            }
                        }
                        Log.d("FetchedQuestions", fetchedQuestions.toString()) // Log the fetched questions
                        _questions.value = fetchedQuestions
                        _isLoading.value = false
                    }
                    .addOnFailureListener { error ->
                        Log.e("Error fetching questions", error.toString())
                        _isLoading.value = false
                    }
            } else {
                _isLoading.value = false
            }
        }
    }



    private fun getRandomIds(n: Int, totalQuestions: Int): List<String> {
        val randomIds = mutableSetOf<String>()
        while (randomIds.size < n) {
            val randomId = (1..totalQuestions).random()
            randomIds.add(randomId.toString())
        }
        return randomIds.toList()
    }

}

File: iOSApp/iOS_QuizApp/iOSQuizApp/QuestionsRepository.swift
//
//  QuestionsRepository.swift
//  
//
//  Created by Edwin Zhang.
//

import FirebaseFirestore
import Combine

class QuestionsRepository: ObservableObject {
    private let db = Firestore.firestore()
    @Published private(set) var questions: [Question] = []
    @Published private(set) var isLoading: Bool = false
    
    init() {
       // fetchQuestions()
    }
    
    func fetchQuestions() {
        isLoading = true
        let infoRef = db.collection("Quiz").document("Coding_questions_info")
        
        DispatchQueue.global(qos: .background).async {
            infoRef.getDocument { infoSnap, error in
                guard let infoSnap = infoSnap, error == nil else {
                    print("Error fetching question info: \(String(describing: error))")
                    DispatchQueue.main.async {
                        self.isLoading = false
                    }
                    return
                }
                
                let totalQuestions = infoSnap.data()?["totalQuestions"] as? Int ?? 0
                let numQuestions = 10
                let randomIds = self.getRandomIds(n: numQuestions, totalQuestions: totalQuestions)
                
                let group = DispatchGroup()
                var fetchedQuestions: [Question] = []
                
                for id in randomIds {
                    group.enter()
                    self.db.collection("Quiz").document(id).collection("Coding_questions").getDocuments { snapshot, error in
                        defer { group.leave() }
                        guard let snapshot = snapshot, error == nil else {
                            print("Error fetching question with ID \(id): \(String(describing: error))")
                            return
                        }
                        
                        for doc in snapshot.documents {
                            if let question = Question.fromFirestoreData(id: id, data: doc.data()) {
                                fetchedQuestions.append(question)
                            }
                        }
                    }
                }
                
                group.notify(queue: .main) {
                    self.questions = fetchedQuestions
                    self.isLoading = false
                }
            }
        }
    }

    
    private func getRandomIds(n: Int, totalQuestions: Int) -> [String] {
        var randomIds = Set<String>()
        while randomIds.count < n {
            let randomId = Int.random(in: 1...totalQuestions)
            randomIds.insert(String(randomId))
        }
        return Array(randomIds)
    }
}
